### PR TITLE
Websocket watcher

### PR DESCRIPTION
**Pull request by capyq**
This is a web socket tools. 
How to use: 
- Add Websocket component and list all ws url. 
- when you need to listen message, use the new hook useListerner(urlKey, callback) 
     - urlkey => the key provide in the component Websecoket
     - callback => function to call when message arrive.

### CODE REVIEW

File: src/components/websocket/Websocket.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
Comment (sBouzols) [inline src/components/websocket/Websocket.tsx:2]: 2024 license header

File: src/components/websocket/Websocket.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
```
Comment (sBouzols) [inline src/components/websocket/Websocket.tsx:7]: Maybe add @author on new files

File: demo/src/app.jsx [hunk 3]
```diff
@@ -888,6 +879,14 @@ function AppContent({ language, onLanguageClick }) {
     );
 }
 
+const WS_URL = 'WS_URL';
+const WS_URL_KEY = 'WS_URL_KEY';
+function AppWSConsumer() {
+    useListener(WS_URL_KEY, (event) => {
+        console.log('WS called UPDATE_FINISHED : ', event);
+    });
```
Comment (sBouzols) [inline demo/src/app.jsx:887]: Where do you simulate a notification event for `WS_URL_KEY` in the demo ?

File: demo/src/app.jsx [hunk 4]
```diff
@@ -912,10 +911,13 @@ function App() {
                 locale={computedLanguage}
                 messages={messages[computedLanguage]}
             >
-                <AppContent
-                    language={language}
-                    onLanguageClick={handleLanguageClick}
-                />
+                <Websocket urls={{ [WS_URL_KEY]: WS_URL }}>
```
Comment (sBouzols) [inline demo/src/app.jsx:914]: What happens if we use the `useListener` without using HOC `Websocket` somewere in the hierarchy ?
Seems like they are linked component/hook.

Comment (capyq) [reply demo/src/app.jsx:914]: yes there are linked as context Providers and useContext are.

Comment (capyq) [reply demo/src/app.jsx:914]: f you use `useContext` without the provider, an error is thrown indicating that the context is not found.

File: demo/src/app.jsx (hunk 3 above)
Comment (capyq) [reply demo/src/app.jsx]: remove from demo.

File: package.json [hunk 5]
```diff
@@ -30,6 +30,7 @@
     "dependencies": {
         "@react-querybuilder/dnd": "^7.2.0",
         "@react-querybuilder/material": "^7.2.0",
+        "@testing-library/react": "^16.0.0",
```
Comment (sBouzols) [inline package.json:33]: in dev dependencies ?

File: src/components/websocket/Websocket.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import ReconnectingWebSocket from 'reconnecting-websocket';
+import { ListenerWS, WSContext, WSContextType } from './contexts/WSContext';
+
+// the delay before we consider the WS truly connected
+const DELAY_BEFORE_WEBSOCKET_CONNECTED = 12000;
+
+type WebsocketProps = { urls: Record<string, string> };
+function Websocket({ urls, children }: PropsWithChildren<WebsocketProps>) {
+    const urlsListenersRef = useRef(
+        Object.keys(urls).reduce((acc, urlKey) => {
```
Comment (jonenst) [inline src/components/websocket/Websocket.tsx:25]: not recomputed if urls change

File: src/components/websocket/Websocket.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+} from 'react';
+import ReconnectingWebSocket from 'reconnecting-websocket';
+import { ListenerWS, WSContext, WSContextType } from './contexts/WSContext';
+
+// the delay before we consider the WS truly connected
+const DELAY_BEFORE_WEBSOCKET_CONNECTED = 12000;
+
+type WebsocketProps = { urls: Record<string, string> };
+function Websocket({ urls, children }: PropsWithChildren<WebsocketProps>) {
+    const urlsListenersRef = useRef(
```
Comment (capyq) [inline src/components/websocket/Websocket.tsx:24]: update ref when url change

File: src/components/websocket/hooks/useListener.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// @author Quentin CAPY
+
+import { useContext, useEffect } from 'react';
+import { WSContext } from '../contexts/WSContext';
+
+export const randomUniqueID = (length: number) =>
```
Comment (capyq) [inline src/components/websocket/hooks/useListener.ts:12]: replace with uuid.v4

**Comment by capyq** [comment]
@sBouzols rebase done.

File: src/components/notifications/hooks/useListenerManager.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    TMessage extends MessageEvent | never
+>(
+    urls: Record<string, string>
+) => {
+    const urlsListenersRef = useRef(
+        Object.keys(urls).reduce((acc, urlKey) => {
+            acc[urlKey] = [];
+            return acc;
+        }, {} as Record<string, TListener[]>)
+    );
```
Comment (TheMaskedTurtle) [inline src/components/notifications/hooks/useListenerManager.ts:22]:
```suggestion
   const urlsListenersRef = useRef<Record<string, TListener[]>>({});
```

**Review by TheMaskedTurtle — APPROVED** [review]

File: src/components/notifications/hooks/useListenerManager.ts (hunk 9 above)
Comment (sBouzols) [reply src/components/notifications/hooks/useListenerManager.ts:22]: fixed

**Pull request merged**